### PR TITLE
Fix right JOINS with parallel replicas

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixes RIGHT / FULL joins in queries with parallel replicas. Now, RIGHT joins can be executed with parallel replicas (right table reading is distributed). FULL joins can't be parallelized among nodes, - executed locally

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 36074938812ec0e279c5aa8cddba590b9acdee70 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71162/36074938812ec0e279c5aa8cddba590b9acdee70/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71162/36074938812ec0e279c5aa8cddba590b9acdee70/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/02967_parallel_replicas_joins_and_analyzer.reference [hunk 1]
```diff
@@ -266,24 +266,13 @@ Expression
...
-              ReadFromMergeTree
-          Expression
-            ReadFromRemoteParallelReplicas
-    Union
-      Expression
         Expression
-          ReadFromMergeTree
-      Expression
-        ReadFromRemoteParallelReplicas
+          ReadFromMemoryStorage
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply tests/queries/0_stateless/02967_parallel_replicas_joins_and_analyzer.reference:270]: No, that's ok.

[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply tests/queries/0_stateless/02967_parallel_replicas_joins_and_analyzer.reference:270]: We can fix this later, if we allow using the left table as a temporary one for GLOBAL JOIN.

**Review by nickitat — APPROVED** [review]
lgtm

File: tests/queries/0_stateless/02967_parallel_replicas_joins_and_analyzer.reference (hunk 1 above)
[the remark this answers is not in the document]
Comment (devcrafter) [reply tests/queries/0_stateless/02967_parallel_replicas_joins_and_analyzer.reference]: Fixed in [d34f11a](https://github.com/ClickHouse/ClickHouse/pull/71162/commits/d34f11a833104a039dafc15eb597c8078a6f13ea)

**Pull request merged**